### DISCUSSION TITLE
private 메서드의 테스트는 어떻게 해야 하나요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#43] [Test]
(주관)할 필요가 없다. private 메서드를 테스트하고 싶다면, 클래스를 분리해야 하는 신호가 아닐지 생각해보자.

**Comment by kunsanglee** [comment]
private 메서드는 그 클래스의 public 또는 protected 메서드를 통해 간접적으로 테스트되어야 합니다. 이러한 관점에서, private 메서드를 직접 테스트할 필요가 있다고 느낀다면, 몇 가지를 고려해볼 필요가 있습니다.

- 리팩토링의 신호: 만약 private 메서드를 직접 테스트하고 싶다면, 이는 클래스가 너무 많은 책임을 가지고 있거나, 코드가 적절히 분리되지 않았을 수 있는 신호입니다. 이 경우, 클래스를 더 작고 관리하기 쉬운 부분으로 분리하는 리팩토링을 고려해볼 수 있습니다. 분리된 각 클래스는 자체적인 public 메서드를 가지며, 이를 통해 간접적으로 private 로직을 테스트할 수 있습니다.

- 테스트 가능한 설계: 소프트웨어 설계의 중요한 목표 중 하나는 테스트 가능성입니다. 클래스나 메서드를 설계할 때, 테스트가 용이하도록 고민하는 것이 중요합니다. 이는 종종 더 좋은 추상화와 책임의 명확한 분리로 이어집니다.

- 행위에 집중: 테스트는 주로 클래스나 메서드의 행위에 집중해야 합니다. 즉, 어떤 입력이 주어졌을 때 기대하는 출력이나 상태 변화가 무엇인지에 초점을 맞추는 것이 중요합니다. 이러한 관점에서 볼 때, private 메서드의 구현 세부 사항보다는 클래스가 외부에 제공하는 인터페이스와 그 행위를 테스트하는 것이 더 중요합니다.

결론적으로, private 메서드의 테스트는 클래스의 public 인터페이스를 통해 간접적으로 이루어져야 합니다. 만약 private 메서드를 직접 테스트하고 싶은 욕구가 있다면, 이는 설계를 재고해볼 필요가 있는 신호일 수 있으며, 코드의 구조나 책임 분리에 대해 다시 생각해볼만 합니다.

**Comment by seunghye218** [comment]
일반적으로 private 메소드에 대한 직접 테스트를 작성할 필요가 없다고 생각합니다. 그 이유는 private 메소드가 클래스의 구현 세부 사항이므로 public 인터페이스가 테스트의 초점이 되어야 한다는 것입니다.

private 메소드는 외부 세계로부터 숨겨진 클래스의 내부 구현 세부정보를 의미합니다. 직접 테스트하면 캡슐화 원칙을 위반하고 이러한 구현 세부 정보가 불필요하게 노출될 수 있습니다.

private 메소드를 직접 테스트하면 테스트와 클래스의 내부 구현 간의 결합이 증가하여 구현이 변경될 때 테스트가 더 취약해지고 깨지기 쉽습니다.
